---
title: "Signac v0.1.0"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

# Signac

Signac is an extension of [Seurat](https://github.com/satijalab/seurat) for the analysis, interpretation, and exploration of single-cell chromatin datasets.

## Features

Signac is currently focused on the analysis of single-cell ATAC-seq data, but new features will be added as experimental methods for measuring other chromatin-based modalities at single-cell resolution become more widespread. Signac is under active development, and the software supports the following features:

* Calculating single-cell QC metrics
* Dimensional reduction, visualization, and clustering
* Identifying cell type-specific peaks
* Visualizing 'pseudo-bulk' coverage tracks
* Integration with single-cell RNA-seq datasets

Please see the Signac vignettes page for examples.

## Installation

To use Signac first install the package dependencies:

```{r}
# GitHub dependency
installed.packages("devtools")
devtools::install_github("thomasp85/patchwork")

# Install bioconductor
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install()

# Tell R to also check bioconductor when installing dependencies
setRepositories(ind=1:2)
```

Finally, install Signac:

```{r}
devtools::install_github("timoast/signac")
```

For information about installing Seurat, see the Seurat [website](https://satijalab.org/seurat/install.html)

It can also be useful (but not essential) to install species-specific packages from bioconductor:

#### Human 

```{r}
BiocManager::install(c('BSgenome.Hsapiens.UCSC.hg19', 'EnsDb.Hsapiens.v75'))
```

#### Mouse

```{r}
BiocManager::install(c('BSgenome.Mmusculus.UCSC.mm10', 'EnsDb.Mmusculus.v79'))
```
