# Set a default value if an object is null
#
# @param lhs An object to set if it's null
# @param rhs The value to provide if x is null
#
# @return rhs if lhs is null, else lhs
#
# @author Hadley Wickham
# @references https://adv-r.hadley.nz/functions.html#missing-arguments
#
`%||%` <- function(lhs, rhs) {
  if (!is.null(x = lhs)) {
    return(lhs)
  } else {
    return(rhs)
  }
}

PlanThreads <- function ()
{
  nthreads <- eval(expr = formals(fun = plan())$workers)
  return(nthreads %||% 1)
}

#' StringToGRanges
#'
#' Convert a genomic coordinate string to a GRanges object
#'
#' @param regions Vector of genomic region strings
#' @param sep Vector of separators to use for genomic string. First element is used to separate chromosome
#' and coordinates, second separator is used to separate start and end coordinates.
#' @return Returns a GRanges object
#' @importFrom GenomicRanges makeGRangesFromDataFrame
#' @importFrom tidyr separate
#' @examples
#' regions <- c('chr1-1-10', 'chr2-12-3121')
#' StringToGRanges(regions = regions)
#'
#' @export
StringToGRanges <- function(regions, sep = c("-", "-")) {
  ranges.df <- data.frame(ranges = regions)
  ranges.df <- separate(
    data = ranges.df,
    col = 'ranges',
    sep = paste0(sep[[1]], "|", sep[[2]]),
    into = c('chr', 'start', 'end')
  )
  granges <- makeGRangesFromDataFrame(df = ranges.df)
  return(granges)
}

#' GRangesToString
#'
#' Convert GRanges object to a vector of strings
#'
#' @param grange A GRanges object
#' @param sep Vector of separators to use for genomic string. First element is used to separate chromosome
#' and coordinates, second separator is used to separate start and end coordinates.
#' @importFrom GenomicRanges seqnames start end
#' @export
GRangesToString <- function(grange, sep = c("-", "-")) {
  regions <- paste0(as.character(seqnames(x = grange)), sep[[1]], start(grange), sep[[2]], end(grange))
  return(regions)
}

#' CalculateCoverages
#'
#' Calculate normalized read coverage per base per cell group
#'
#' @param reads Dataframe containing reads for a region, generated by \code{GetReadsInRegion}
#' @param window Smoothing window to use
#'
#' @importFrom zoo rollapply
#' @importFrom dplyr bind_rows
#' @export
CalculateCoverages <- function(
  reads,
  window = 100
) {
  cells.per.group <- table(reads$group)
  lut <- as.vector(x = cells.per.group)
  names(lut) <- names(x = cells.per.group)
  expanded <- suppressWarnings(bind_rows(lapply(X = 1:nrow(reads), FUN = function(x) {
    interval = reads[x, 'start']:reads[x, 'stop']
    return(data.frame(position = interval, value = 1, cell = reads[x, 'cell'], group = reads[x, 'group']))
  })))
  expanded$norm.value <- expanded$value / as.vector(x = lut[as.character(x = expanded$group)])
  expanded$coverage <- rollapply(data = expanded$norm.value, width = window, FUN = mean, align = 'center', fill = NA)
  return(expanded)
}

#' ConstructBinMatrix
#'
#' Construct a bin x cell matrix from a fragments file
#'
#' @param fragments Path to tabix-indexed fragments file
#' @param genome A GRanges object for the genome
#' @param cells Vector of cells to include. If NULL, include all cells found
#' in the fragments file
#' @param binsize Size of the genome bins to use
#' @param chunk Number of chunks to use when processing the fragments file. Fewer chunks may enable faster processing,
#'  but will use more memory.
#' @param sep Vector of separators to use for genomic string. First element is used to separate chromosome
#' and coordinates, second separator is used to separate start and end coordinates.
#' @param verbose Display messages
#'
#' @importFrom GenomicRanges tileGenome
#' @importFrom GenomeInfoDb seqlengths
#' @importFrom future.apply future_lapply
#' @importFrom pbapply pblapply
#' @importFrom Matrix sparseMatrix
#'
#' @export
ConstructBinMatrix <- function(
  fragments,
  genome,
  cells = NULL,
  binsize = 5000,
  chunk = 100,
  sep = c('-', '-'),
  verbose = TRUE
) {
  tiles <- tileGenome(seqlengths = seqlengths(genome), tilewidth = binsize, cut.last.tile.in.chrom = TRUE)
  chunks <- unlist(x = tileGenome(seqlengths = seqlengths(genome), ntile = chunk))
  stringchunks <- GRangesToString(grange = chunks, sep = c(':', '-'))
  if (verbose) {
    message('Extracting reads overlapping genome bins')
  }
  if (PlanThreads() > 1) {
    mylapply <- future_lapply
  } else {
    mylapply <- ifelse(test = verbose, yes = pblapply, no = lapply)
  }
  cells.in.regions <- mylapply(
    X = stringchunks,
    FUN = GetCellsInBin,
    fragments = fragments,
    cells = cells,
    bins = tiles
  )
  if (verbose) {
    message("Constructing matrix")
  }
  bin.regions <- unlist(x = lapply(X = cells.in.regions, FUN = `[[`, 1))
  cell.vector <- unlist(x = lapply(X = cells.in.regions, FUN = `[[`, 2))
  all.cells <- unique(x = cell.vector)
  cell.lookup <- seq_along(along.with = all.cells)
  names(x = cell.lookup) <- all.cells
  binmat <- sparseMatrix(
    i = bin.regions,
    j = cell.lookup[cell.vector],
    x = rep(x = 1, length(x = cell.vector))
  )
  binmat <- as(Class = 'dgCMatrix', object = binmat)
  tilestring <- GRangesToString(grange = tiles, sep = sep)
  bin.range <- 1:max(bin.regions)
  rownames(binmat) <- tilestring[bin.range]
  colnames(binmat) <- names(x = cell.lookup)
  return(binmat)
}

#' Extend
#'
#' Resize GenomicRanges upstream and or downstream.
#' From \url{https://support.bioconductor.org/p/78652/}
#'
#' @param x A range
#' @param upstream Length to extend upstream
#' @param downstream Length to extend downstream
#'
#' @importFrom GenomicRanges strand start end trim
#' @importFrom IRanges ranges IRanges
#' @export
Extend <- function(x, upstream = 0, downstream = 0) {
  if (any(strand(x = x) == "*")) {
    warning("'*' ranges were treated as '+'")
  }
  on_plus <- strand(x = x) == "+" | strand(x = x) == "*"
  new_start <- start(x = x) - ifelse(test = on_plus, yes = upstream, no = downstream)
  new_end <- end(x = x) + ifelse(test = on_plus, yes = downstream, no = upstream)
  ranges(x = x) <- IRanges(start = new_start, end = new_end)
  x <- trim(x = x)
  return(x)
}

#' FilterFragments
#'
#' Remove cells from a fragments file that are not present in a given list of cells.
#'
#' @param fragment.path Path to a tabix-indexed fragments file
#' @param cells A vector of cells to retain
#' @param output.path Name and path for output tabix file. A tabix index file will also be created in the same location, with
#' the .tbi file extension.
#' @param chunk Number of chunks to use when processing the fragments. Fewer chunks may be faster, but will use more memory.
#' @param verbose Display messages
#'
#' @importFrom seqminer tabix.read.table tabix.createIndex
#' @export
FilterFragments <- function(
  fragment.path,
  cells,
  output.path,
  chunk,
  verbose
) {
  if (verbose) {
    message("Filtering cell from fragments file. Retaining ", length(cells), " cells")
  }
  #
  if (verbose) {
    message("Completed")
  }
}

#' GetCellsInBin
#'
#' Extract cell names containing reads mapped within a given genomic region
#'
#' @param fragments Tabix-indexed fragments file
#' @param region A string giving the region to extract from the fragments file
#' @param bins A GRanges object containing the full set of genome bins
#' @param cells Vector of cells to include in output. If NULL, include all cells
#'
#' @importFrom seqminer tabix.read.table
#' @importFrom GenomicRanges makeGRangesFromDataFrame findOverlaps
#' @export
GetCellsInBin <- function(fragments, region, bins, cells = NULL) {
  bin.reads <- tabix.read.table(
    tabixFile = fragments,
    tabixRange = region
  )
  if (nrow(x = bin.reads) > 0) {
    if (!is.null(x = cells)) {
      bin.reads <- bin.reads[bin.reads$V4 %in% cells, ]
    }
    if (nrow(x = bin.reads) > 0) {
      colnames(bin.reads) <- c('chrom', 'start', 'end', 'cell', 'count')
      gr.reads <- makeGRangesFromDataFrame(df = bin.reads)
      overlaps <- findOverlaps(query = gr.reads, subject = bins, select = 'first')
      return(list(bin = overlaps, cell = bin.reads$cell))
    } else {
      return(NULL)
    }
  } else {
    return(NULL)
  }
}

#' GetReadsInRegion
#'
#' Extract reads for each cell within a given genomic region or set of regions
#'
#' @param object A Seurat object
#' @param region A genomic region, specified as a string in the format 'chr:start-end'. Can be a vector of regions.
#' @param assay Name of assay to use
#' @param fragment.path Path to indexed fragment file
#' @param group.by Cell grouping information to add
#' @param cells Cells to include. Default is all cells present in the object.
#' @param verbose Display messages
#'
#' @importFrom seqminer tabix.read.table
#' @importFrom future plan
#'
#' @return Returns a data frame
#' @export
GetReadsInRegion <- function(
  object,
  region,
  assay = NULL,
  fragment.path = NULL,
  group.by = NULL,
  cells = NULL,
  verbose = TRUE
) {
  assay <- assay %||% DefaultAssay(object = object)
  if (is.null(group.by)) {
    group.by <- Idents(object = object)
  } else {
    meta.data <- object[[]]
    group.by <- meta.data[[group.by]]
    names(group.by) <- rownames(x = meta.data)
  }
  if (is.null(x = fragment.path)) {
    tools <- slot(object = object, name = 'tools')
    if ('fragments' %in% names(x = tools)) {
      if (assay %in% names(x = tools$fragments)) {
        fragment.path <- tools$fragments[[assay]]
      } else {
        stop('Fragment file not supplied for the requested assay')
      }
    } else {
      stop('Fragment file not set. Run SetFragments to set the fragment file path.')
    }
  } else if (!(all(file.exists(fragment.path, paste0(fragment.path, '.tbi'))))) {
    stop('Requested file does not exist or is not indexed')
  }
  if (verbose) {
    message('Extracting reads in requested region')
  }
  reads <- tabix.read.table(tabixFile = fragment.path, tabixRange = region)
  colnames(reads) <- c('chrom', 'start', 'stop', 'cell', 'reads')
  reads <- reads[reads$cell %in% names(group.by), ]
  if (!is.null(x = cells)) {
    reads <- reads[reads$cell %in% cells, ]
  }
  if (nrow(reads) == 0) {
    stop('No cells present in the requested region')
  }
  reads$length <- reads$stop - reads$start
  reads$group <- group.by[reads$cell]
  return(reads)
}

#' CountsInRegion
#'
#' Count reads per cell overlapping a given set of regions
#'
#' @param object A Seurat object
#' @param assay Name of assay in the object to use
#' @param regions A GRanges object
#' @param sep Separator to use when extracting genomic coordinates from the Seurat object
#' @param ... Additional arguments passed to \code{findOverlaps}
#'
#' @importFrom IRanges findOverlaps
#' @importFrom S4Vectors queryHits
#' @importFrom Matrix colSums
#'
#' @export
CountsInRegion <- function(
  object,
  assay,
  regions,
  sep = c("-", '-'),
  ...
) {
  obj.regions <- rownames(x = object[[assay]])
  obj.granges <- StringToGRanges(regions = obj.regions, sep = sep)
  overlaps <- findOverlaps(query = obj.granges, subject = regions, ...)
  hit.regions <- GRangesToString(grange = obj.granges[queryHits(x = overlaps)], sep = sep)
  data.matrix <- GetAssayData(object = object, assay = assay, slot = 'counts')[hit.regions, ]
  return(colSums(data.matrix))
}

#' FractionCountsInRegion
#'
#' Find the fraction of counts per cell that overlap a given set of genomic ranges
#'
#' @param object A Seurat object
#' @param assay Name of assay to use
#' @param regions A GRanges object containing a set of genomic regions
#' @param sep The separator used to separate genomic coordinate information in the assay feature names
#' @param ... Additional arguments passed to \code{\link{CountsInRegion}}
#'
#' @export
FractionCountsInRegion <- function(
  object,
  assay,
  regions,
  sep = c("-", "-"),
  ...
) {
  reads.in.region <- CountsInRegion(
    object = object,
    regions = regions,
    assay = assay,
    sep = sep,
    ...
  )
  total.reads <- colSums(GetAssayData(object = object, assay = assay, slot = 'counts'))
  return(reads.in.region / total.reads)
}

#' Set the fragments file path for creating plots
#'
#' Give path of indexed fragments file that goes with data in the object.
#' Checks for a valid path and an index file with the same name (.tbi) at the same path.
#' Stores the path under the tools slot for access by visualization functions.
#' One fragments file can be stored for each assay.
#'
#' @param object A Seurat object
#' @param file Path to indexed fragment file. See \url{https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/fragments}
#' @param assay Assay used to generate the fragments. If NULL, use the active assay.
#'
#' @export
#'
SetFragments <- function(
  object,
  file,
  assay = NULL
) {
  assay <- assay %||% DefaultAssay(object = object)
  if (!(assay %in% names(x = slot(object = object, name = 'assays')))) {
    stop('Requested assay not present in object')
  }
  index.file <- paste0(file, '.tbi')
  if (all(file.exists(file, index.file))) {
    file <- normalizePath(path = file)
    current.tools <- slot(object = object, name = 'tools')
    current.tools$fragments[[assay]] <- file
    slot(object = object, name = 'tools') <- current.tools
    return(object)
  } else {
    stop('Requested file does not exist or is not indexed')
  }
}
